### DISCUSSION TITLE
Performance: Object sleep API

### DISCUSSION THREAD

**Discussion by parasyte** [melonjs/melonJS#1023] [Ideas]
Following [discussion on the forum](https://groups.google.com/forum/?fromgroups=#!topic/melonjs/YiXmj51ogqU), it sounds like a good idea to implement an object sleep mechanism to save CPU cycles for objects that have nothing to do.

The way I imagine it is having two pools for game objects; the `gameObjects` array in core.js, and a new array for `sleepingObjects`. Calling `me.game.sleep(obj)` will remove `obj` from `gameObjects`, and push it into `sleepingObjects`. And `me.game.wakeup(obj)` will do the opposite.

A sleeping object will never be iterated in the update or draw loops, saving precious CPU time.

Further, we can enable a flag like `me.sys.enableAutoSleep` to put objects to sleep when they leave the viewport. Deciding on a good method of waking up a sleeping object without polling will be crucial for this flag to be useful. I suppose worst-case, polling infrequently (once per second?) could work.

**Comment by obiot** [comment]
I didn't really had the time to check the forum since yesterday since I was travelling, but this seems to be a good idea !

Not sure though  however about the automatic feature, as both putting them to sleep or awake  automatically would effectively could be quite CPU consuming .

**Comment by obiot** [comment]
@parasyte though it's not really related, I also wanted to look at replacing the current object array by a hash table (using the object GUID for key), as it should improve performances as well when parsing and accessing the object pool/list

**Comment by obiot** [comment]
Improvements when doing lookup : http://jsperf.com/indexof-vs-array-hash

**Comment by parasyte** [comment]
I had the same idea. See also a similar test in #192 that shows how faking a remove operation is faster than both `Array.pop()` and `Object.delete()`.

**Comment by parasyte** [comment]
@obiot: auto-sleep is easy. Just plug it in right here: https://github.com/melonjs/melonJS/blob/master/src/core.js#L1422 But auto-wakeup is _really hard_.

**Comment by obiot** [comment]
Yes I was actually more concerned about the auto-wake up :)

**Comment by parasyte** [comment]
Reminder: Implementing auto-sleep through callbacks is the right way to do this. A "viewport leave" callback is very easy, as we know. With the new collision detection in #103, we will have a spatial grid that already contains objects... Seems like this can be reused for auto-wakeup, too! When a new spatial grid cell enters the viewport, fire the "wakeup" callback on all objects in that cell.

Wow ... that seems almost _too_ easy to be true!

**Comment by obiot** [comment]
something interesting i just spotted in the chipmunk API :
http://chipmunk-physics.net/release/ChipmunkLatest-Docs/#cpSpace-Sleeping

it's also possible to disable object collision check for idle object, and specify a `sleepTimeThreshold`, might be also a nice addition if we do add this sleeping API.
